src/machine/machine_stm32f407.go

```diff
@@ -197,24 +130,24 @@ func (uart UART) Configure(config UARTConfig) {
 	  | 115200   | 0x16D  |
 	  +----------+--------+
 	*/
-	stm32.USART2.BRR.Set(0x16c)
+	uart.Bus.BRR.Set(0x16c)
 
 	// Enable USART2 port.
-	stm32.USART2.CR1.Set(stm32.USART_CR1_TE | stm32.USART_CR1_RE | stm32.USART_CR1_RXNEIE | stm32.USART_CR1_UE)
+	uart.Bus.CR1.Set(stm32.USART_CR1_TE | stm32.USART_CR1_RE | stm32.USART_CR1_RXNEIE | stm32.USART_CR1_UE)
 
-	// Enable RX IRQ.
+	// Enable RX IRQ. TODO: pick the right IRQ_xxx for the bus and the uart
```

Comment:
You can take a look here for how this can be done:
https://github.com/tinygo-org/tinygo/blob/1d913a62bc0ceef841953058c2c6fb1dd2a40f6a/src/machine/board_circuitplay_express_baremetal.go#L10-L21
This pattern is very well optimized by the compiler.

Reply:
I was hoping to do something like:
```
var (
  UART1 = UART{
 		Buffer: NewRingBuffer(), 
 		Bus:     stm32.USART2, 
 		IRQ:     stm32.IRQ_USART2 , 
 	} 
) 

...

   intr := interrupt.New(uart.IRQ, func(interrupt.Interrupt) {
		uart.Receive(byte((uart.Bus.DR.Get() & 0xFF)))
	})
```

This way the `Configure()` call didn't need to know about specific bindings of IRQs/Buses to `uart` instances.

Reply:
Yeah I was hoping for that but both have a problem:

The first is not currently supported by the compiler. The `interrupt.New` function is a compiler intrinsic and will only accept constant parameters (both for the interrupt number and for the handler that you pass to it). In this case `uart.IRQ` is not constant. The problem with non-constant parameters is that it is incredibly hard (if not impossible) to optimize non-constant IRQ numbers.

The second is sadly not allowed by the Go specification because of a circular dependency. I really, really wished it was possible to write an interrupt handler like that because it looks so nice. The example I gave above works around that with an `init` function but is basically the same idea.